### PR TITLE
Fix merging of single value headers

### PR DESCRIPTION
**Pull request by dagnir**
## Description
Use the last seen HTTP/1.1 header value for headers defined to only appear once in an HTTP message instead of merging them all into a list. The order in which header values are inspected is: headers set on the request, overridden headers set on the client, then finally overridden headers set on an SDK request. See https://tools.ietf.org/html/rfc2616#section-4.2 for more information.

## Motivation and Context
Bug fix.

## Testing
`mvn clean install`. Will run integ tests.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: utils/src/main/java/software/amazon/awssdk/utils/http/SdkHttpUtils.java [hunk 1]
```diff
@@ -297,4 +307,12 @@ public static String appendUri(String baseUri, String path) {
     public static Optional<String> firstMatchingHeader(Map<String, List<String>> headers, String header) {
         return allMatchingHeaders(headers, header).findFirst();
     }
+
+    public static Set<String> singleHeaders() {
```
Comment (spfink) [inline utils/src/main/java/software/amazon/awssdk/utils/http/SdkHttpUtils.java]: Is this needed?

Comment (dagnir) [reply utils/src/main/java/software/amazon/awssdk/utils/http/SdkHttpUtils.java]: Not really needed, I just needed to able access the list from the tests, but I can just copy them there.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java [hunk 2]
```diff
@@ -52,7 +54,15 @@ public MergeCustomHeadersStage(HttpClientDependencies dependencies) {
     private final Map<String, List<String>> mergeHeaders(Map<String, List<String>>... headers) {
         Map<String, List<String>> result = new LinkedHashMap<>();
         for (Map<String, List<String>> header : headers) {
-            header.forEach((k, v) -> result.computeIfAbsent(k, ignored -> new ArrayList<>()).addAll(v));
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java]: We're now introducing an implicit 'priority' in the sense that the last header 'wins'. Can we add a comment about that?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java [hunk 3]
```diff
@@ -52,7 +54,15 @@ public MergeCustomHeadersStage(HttpClientDependencies dependencies) {
...
+            header.forEach((headerName, headerValues) -> {
+                List<String> resultHeaderValues = result.computeIfAbsent(headerName, ignored -> new ArrayList<>());
+                if (SdkHttpUtils.isSingleHeader(headerName)) {
+                    resultHeaderValues.clear();
+                    resultHeaderValues.add(CollectionUtils.lastElement(headerValues));
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java]: Should we disallow (or log) when there are multiple values specified in the same headerValues for a single-header?

It's a bit weird if I add multiple content-types at the request level, but it just takes the last one.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java (hunk 2 above)
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java]: Sure, i'll add a note on the setters in the Client/Request override configuration.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java (hunk 3 above)
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java]: I went back and forth with this. Alternatively what do you think about keeping all the values in the current list, but not merging them with the previously seen list?

e.g.

"FooHeader" set by marshaller -> "a"
"FooHeader" override on client -> ["b", "c"]
"FooHeader" override on request -> ["d", "e", "f"]

"FooHeader" on the wire -> ["d", "e", "f"]

Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MergeCustomHeadersStage.java]: After discussing offline, went with the above solution.

**Pull request merged**